test/gateway/eip20_gateway/progress_unstake_with_proof.js

```diff
@@ -0,0 +1,511 @@
+
+const NullAddress = Utils.NULL_ADDRESS;
+const ZeroBytes = Utils.ZERO_BYTES32;
+
+async function setStateRoot(gateway) {
+  let blockHeight = new BN(StubData.co_gateway.redeem.proof_data.block_number, 16);
+  let storageRoot = StubData.co_gateway.redeem.proof_data.storageHash;
+  await gateway.setStorageRoot(blockHeight, storageRoot);
+  return blockHeight;
+}
+
+contract('EIP20Gateway.progressUnstakeWithProof()', function (accounts) {
+
+  let gateway, mockToken, baseToken, unstakeRequest, unstakeMessage,
+    stakeVaultAddress;
+  let bountyAmount = new BN(StubData.gateway.constructor.bountyAmount, 16);
+  let MessageStatusEnum = messageBus.MessageStatusEnum;
+  let redeemRequest = StubData.co_gateway.redeem.params;
+
+  beforeEach(async function () {
+    redeemRequest = StubData.co_gateway.redeem.params;
+    mockToken = await MockToken.new({from: accounts[0]});
+    baseToken = await MockToken.new({from: accounts[0]});
+
+    let organizationAddress = accounts[3];
+    let coreAddress = accounts[5];
+    gateway = await Gateway.new(
+      mockToken.address,
+      baseToken.address,
+      coreAddress,
+      bountyAmount,
+      organizationAddress,
+      NullAddress, //burner
+    );
+
+    unstakeRequest = {
+      beneficiary: redeemRequest.beneficiary,
+      amount: new BN(redeemRequest.amount, 16),
+    };
+
+    let redeemIntentHash = cogatewayUtils.hashRedeemIntent(
+      unstakeRequest.amount,
+      unstakeRequest.beneficiary,
+      StubData.contracts.coGateway
+    );
+
+    unstakeMessage = {
+      intentHash: redeemIntentHash,
+      nonce: new BN(redeemRequest.nonce, 16),
+      gasPrice: new BN(redeemRequest.gasPrice, 16),
+      gasLimit: new BN(redeemRequest.gasLimit, 16),
+      unstakeAccount: redeemRequest.redeemer,
+      hashLock: redeemRequest.hashLock,
+      unlockSecret: redeemRequest.unlockSecret,
+    };
+
+    stakeVaultAddress = await setup(unstakeMessage, gateway, unstakeRequest, stakeVaultAddress, mockToken, accounts);
+
+  });
+
+  it('should unstake with correct parameters', async function () {
+
+    await gateway.setInboxStatus(unstakeMessage.messageHash, MessageStatusEnum.Declared);
+    let blockHeight = await setStateRoot(gateway);
+    let storageProof = StubData.co_gateway.redeem.proof_data.storageProof[0].serializedProof;
+
+    let tx = await gateway.progressUnstakeWithProof(
+      unstakeMessage.messageHash,
+      storageProof,
+      blockHeight,
+      MessageStatusEnum.Declared
+    );
+
+    assert.strictEqual(
+      tx.receipt.status,
+      true,
+      'Transaction should success.',
+    );
+
+  });
+
+  it('should return correct parameters', async function () {
+
+    await gateway.setInboxStatus(unstakeMessage.messageHash, MessageStatusEnum.Declared);
+    let blockHeight = await setStateRoot(gateway);
+    let storageProof = StubData.co_gateway.redeem.proof_data.storageProof[0].serializedProof;
+
+    let response = await gateway.progressUnstakeWithProof.call(
```

Comment:
General question: where is `progressUnstakeWithProof()` called, why is it `public` and not `external` in the smart contract?

Also: is it expected to be called by another contract that uses the return values? Do the return values generally make sense/improve the code?

Reply:
Good catch, I changed it to 'external'. 
Return parameters are added everywhere. This gives the flexiblity of consuming contracts from other contracts. Though,  Kit is the only consumer of protocol currently, but this may change in future.

Reply:
ok